edgelet/edgelet-http/src/lib.rs

```diff
@@ -236,14 +269,18 @@ where
                     let service = PidService::new(pid, srv);
                     protocol
                         .serve_connection(socket, service)
-                        .then(move |result| match result {
+                        .then(move |result| {
+                            if let Some((end_future, _)) = end_future {
+                                end_future.fetch_add(1, Ordering::AcqRel);
+                            }
```

Comment:
nit: this part looks a bit confusing to me @_@... Would you mind explaining what this `end_future` is?

Reply:
just a copy of             let end_future = lock.clone();
Is there a better way than to create an intermediary variable?

Reply:
Just to be clear, the goal is just to make sure that end_future (aka the lock) increments on serve_connection? If so, this looks fine, I just had to scroll up to see what end_future was and if it was the lock you made... maybe a name change from `end_future` to something else might be clearer? If think it is fine, then feel free to resolve :)

Reply:
There's a pattern that Denis had introduced to me that helps.

Instead of:

```rust
let foo = ...;
let foo2 = foo.clone();
let foo3 = foo.clone();
bar(|| {
    do_something_with(foo);
})
.baz(|| {
    do_something_with(foo2);
})
.quux(|| {
    do_something_with(foo3);
})
```

you do:

```rust
let foo = ...;
bar({
    let foo = foo.clone();
    || {
        do_something_with(foo);
    }
})
.baz({
    let foo = foo.clone();
    || {
        do_something_with(foo);
    }
})
.quux(|| {
    do_something_with(foo);
})
```

ie instead of passing a closure directly, you pass a block expr that evaluates to the closure. Since it's a block, it can contain statements before the closure expr, so you can write statements to create all the captures for the closure there. Not only are the captures closer to the closure that captures them, but you don't have to come up with creative unique names because you can just shadow the original binding.

Eg:

- https://github.com/Azure/iotedge/blob/ec491d799711300d098c31b63ce86aba2570aa3c/edgelet/edgelet-docker/src/runtime.rs#L132-L143

- https://github.com/Azure/iotedge/blob/ec491d799711300d098c31b63ce86aba2570aa3c/edgelet/edgelet-http-workload/src/server/cert/mod.rs#L87-L98

Reply:
ahhh! wish I knew that before. Thanks! I reworked it.